core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -1069,25 +1084,56 @@ private[spark] class TaskSetManager(
    * Check if the task associated with the given tid has past the time threshold and should be
    * speculative run.
    */
-  private def checkAndSubmitSpeculatableTask(
-      tid: Long,
+  private def checkAndSubmitSpeculatableTasks(
       currentTimeMillis: Long,
-      threshold: Double): Boolean = {
-    val info = taskInfos(tid)
-    val index = info.index
-    if (!successful(index) && copiesRunning(index) == 1 &&
-        info.timeRunning(currentTimeMillis) > threshold && !speculatableTasks.contains(index)) {
-      addPendingTask(index, speculatable = true)
-      logInfo(
-        ("Marking task %d in stage %s (on %s) as speculatable because it ran more" +
-          " than %.0f ms(%d speculatable tasks in this taskset now)")
-          .format(index, taskSet.id, info.host, threshold, speculatableTasks.size + 1))
-      speculatableTasks += index
-      sched.dagScheduler.speculativeTaskSubmitted(tasks(index))
-      true
-    } else {
-      false
+      threshold: Double,
+      numSuccessfulTasks: Int,
+      customizedThreshold: Boolean = false): Boolean = {
+    var foundTasksResult = false
+    for (tid <- runningTasksSet) {
+      val info = taskInfos(tid)
+      val index = info.index
+      if (!successful(index) && copiesRunning(index) == 1 && !speculatableTasks.contains(index)) {
+        val runtimeMs = info.timeRunning(currentTimeMillis)
+
+        def checkMaySpeculate(): Boolean = {
+          if (customizedThreshold || taskProcessRateCalculator.isEmpty) {
+            true
+          } else {
+            val longTimeTask = runtimeMs > efficientTaskDurationFactor * threshold
```

Comment:
This seems to strict the speculatable condition of a task. Previously, a task can be speculate as long as the `runtimeMs > speculationMultiplier * medianDuration`, but now it has to be `runtimeMs > efficientTaskDurationFactor * speculationMultiplier * medianDuration`. And you can't fallback to the original behaviour even if `spark.speculation.efficiency.enabled` is disabled, while it's supposed to be a global control flag for the whole feature.
Besides, with my understanding, should the condition be like `(runtimeMs > speculationMultiplier * medianDuration) && taskProcessRateCalculator.exists(_.isInefficient(tid, runtimeMs, info))`?

Reply:
When  the  `spark.speculation.efficiency.enabled` is disabled,    the `taskProcessRateCalculator.isEmpty` will be true  and the `checkMaySpeculate()` will  always return true.